### ISSUE TITLE
Parser expects `eslint.lintText`'s `options.filePath` to be present in the filesystem even when used programmatically.

### ISSUE THREAD

**Issue by yudai-nkt** [typescript-eslint/typescript-eslint#4360]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

`@typescript-eslint/parser` requires the value for [`eslint.lintText`](https://eslint.org/docs/developer-guide/nodejs-api#-eslintlinttextcode-options)'s `options.filePath` to be actually present in the filesystem. Is this intended?

**Repro**

```JSON
{
  "overrides": [
    {
      "files": ["*.ts"],
      "parser": "@typescript-eslint/parser",
      "parserOptions": {
        "project": "./tsconfig.json"
      }
    }
  ]
}
```

```javascript
// NOTE: This is an issue with using ESLint programmatically.
import { ESLint } from "eslint";

const eslint = new ESLint();
const [{ messages }] = await eslint.lintText("", { filePath: "index.ts" });
console.log(messages);
```

Save the above JavaScript code as `test.mjs` and run `node test.mjs`. The content of `tsconfig.json` is what `npx tsc --init` generates by default.

**Expected Result**

Linting succeeds and reports no error.

**Actual Result**

A parsing error occurs:

```console
$ node test.mjs
[
  {
    ruleId: null,
    fatal: true,
    severity: 2,
    message: 'Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.\n' +
      'The file does not match your project config: index.ts.\n' +
      'The file must be included in at least one of the projects provided.',
    line: undefined,
    column: undefined
  }
]
```

If I make `index.ts`, the error disappears:
```console
$ touch index.ts && node test.mjs
[]
```
**Additional Info**

I thought this is an issue with the parser because the following patch also eliminates the error, but correct me if I'm wrong.
```diff
-const [{ messages }] = await eslint.lintText("", { filePath: "index.ts" });
+const [{ messages }] = await eslint.lintText("", { filePath: "index.js" });
```

I just thought `{ filePath: "index.ts"}` matches the ESLint's `overrides` option and enables the TypeScript parser. I personally didn't expect that `index.ts` needs to be created and this behavior took me some time to troubleshoot my failing tests.

**Versions**

| package                     | version |
| --------------------------- | ------- |
| `@typescript-eslint/parser` | `5.8.1` |
| `TypeScript`                | `4.5.4` |
| `ESLint`                    | `8.5.0` |
| `node`                      | `16.13.1` |

**Comment by bradzacher** [comment]
This is working as expected.

The ESLint API doesn't tell the parser how it was invoked, just that it was invoked. So all our parser knows is "ESLint wants you to parse this text which is the contents of a file called `index.ts`".

TypeScript operates on files on disk. There are ways to make it work on in-memory files, but without a mechanism to tell us that you're operating on "fake" files, we can only assume the 99.9999% usecase which is files on disk.

You can use `parserOptions.program` to pass a custom TS program built with an in-memory filesystem if you want to use it that way.
